sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/JoinEstimation.scala

```diff
@@ -0,0 +1,314 @@
+
+  /**
+   * Estimate output size and number of rows after a join operator, and update output column stats.
+   */
+  def doEstimate(): Option[Statistics] = join match {
+    case _ if !rowCountsExist(conf, join.left, join.right) =>
+      None
+
+    case ExtractEquiJoinKeys(joinType, leftKeys, rightKeys, condition, left, right) =>
+      // 1. Compute join selectivity
+      val joinKeyPairs = extractJoinKeys(leftKeys, rightKeys)
+      val selectivity = joinSelectivity(joinKeyPairs, leftStats, rightStats)
+
+      // 2. Estimate the number of output rows
+      val leftRows = leftStats.rowCount.get
+      val rightRows = rightStats.rowCount.get
+      val innerRows = ceil(BigDecimal(leftRows * rightRows) * selectivity)
+
+      // Make sure outputRows won't be too small based on join type.
+      val outputRows = joinType match {
+        case LeftOuter =>
+          // All rows from left side should be in the result.
+          leftRows.max(innerRows)
+        case RightOuter =>
+          // All rows from right side should be in the result.
+          rightRows.max(innerRows)
+        case FullOuter =>
+          // Simulate full outer join as obtaining the number of elements in the union of two
+          // finite sets: A \cup B = A + B - A \cap B => A FOJ B = A + B - A IJ B.
+          // But the "inner join" part can be much larger than A \cap B, making the simulated
+          // result much smaller. To prevent this, we choose the larger one between the simulated
+          // part and the inner part.
+          (leftRows + rightRows - innerRows).max(innerRows)
+        case _ =>
+          // Don't change for inner or cross join
+          innerRows
+      }
+
+      // 3. Update statistics based on the output of join
+      val intersectedStats = if (selectivity == 0) {
+        AttributeMap[ColumnStat](Nil)
+      } else {
+        updateIntersectedStats(joinKeyPairs, leftStats, rightStats)
+      }
+      val inputAttrStats = AttributeMap(
+        leftStats.attributeStats.toSeq ++ rightStats.attributeStats.toSeq)
+      val attributesWithStat = join.output.filter(a => inputAttrStats.contains(a))
+      val (fromLeft, fromRight) = attributesWithStat.partition(join.left.outputSet.contains(_))
+      val outputStats: Map[Attribute, ColumnStat] = join.joinType match {
+        case LeftOuter =>
+          // Don't update column stats for attributes from left side.
+          fromLeft.map(a => (a, inputAttrStats(a))).toMap ++
+            updateAttrStats(outputRows, fromRight, inputAttrStats, intersectedStats)
+        case RightOuter =>
+          // Don't update column stats for attributes from right side.
+          updateAttrStats(outputRows, fromLeft, inputAttrStats, intersectedStats) ++
+            fromRight.map(a => (a, inputAttrStats(a))).toMap
+        case FullOuter =>
+          // Don't update column stats for attributes from both sides.
+          attributesWithStat.map(a => (a, inputAttrStats(a))).toMap
```

Comment:
Yes. It's difficult to get a reasonable nullCount estimation after outer joins. So in filter, we only use nullCount to estimate if the child is a table.